### PR TITLE
Fix the handling of PCRE flags.

### PR DESCRIPTION
**Pull request by hvdijk**
This removes the `AST_EXPR_FLAGS` node type. This node type was used to indicate that the following nodes should be interpreted according to its flags, but "the following" was implemented as "the following children of its parent", rather than "the following elements in the original syntax until a closing parenthesis". "The following" was affected by internal re-ordering of nodes in alternatives, and "until a closing parenthesis" could not be determined anyway as non-capturing groups do not get their own nodes.

This commit removes the `AST_EXPR_FLAGS` node entirely and instead modifies the parser to pass the current `re_flags` to all AST construction functions and stores it in a new field in `struct ast_expr`. The compilation in `src/libre/ast_compile.c` is updated to look at this field to find whether `RE_ICASE` is enabled (the only flag currently implemented).

This commit updates the printing of ASTs as trees, dotfiles and PCREs to reconstruct the flags. As they are no longer represented as nodes, they do not get printed as nodes.

This fixes the option handling issues of #200.

Does this look like an okay approach, or would something completely different be better?

### CODE REVIEW

File: src/libre/parser.act [hunk 1]
```diff
@@ -659,28 +654,28 @@
 	 */
 
 	<ast-make-empty>: () -> (node :ast_expr) = @{
-		@node = ast_make_expr_empty();
+		@node = ast_make_expr_empty(*flags);
```
Comment (katef) [inline src/libre/parser.act:657]: Noting to myself that `*flags` comes from a `%persistents%` section

File: src/libre/parser.act [hunk 2]
```diff
@@ -90,11 +90,6 @@
 	#include "../comp.h"
 	#include "../../class.h"
 
-	struct flags {
-		enum re_flags flags;
-		struct flags *parent;
-	};
-
```
Comment (katef) [inline src/libre/parser.act:97]: So if we're setting these globally, and don't keep a stack or a tree, what's happening for cases like this?
```
; echo abcd | pcregrep '((?i:ABC))d'
abcd
; echo abcd | pcregrep '((?i:ABC))D'
;
```

I think we still need the idea of scope there because the flags do get restored outside of the capturing group.

Answer: I see you're saving and restoring the flags in pcre's parser.sid. That makes sense. So we're using the parser's call stack as our mechanism for scope.

This behaviour seems to apply to non-capturing groups, too. But we no longer need an AST node to hold that, because these are now set for every node. Okay! Great.

**Comment by katef** [comment]
AST now:

image

**Comment by katef** [comment]
The force-push above also removes `struct flags` entirely (with its `.pos` and `.neg` fields), leaving us now with just:
```diff
-	typedef struct flags *flags;
+	typedef enum re_flags * flags;
```
So now all flags are phrased in the positive in the AST, the idea of negative flags is now a detail of syntax only.

**Review by katef — APPROVED** [review]

**Comment by hvdijk** [comment]
I found that with this PR, the ast printing of `\` and `|` was broken. These were not handled in `dot_escputc_html_record`, but it did not cause problems before as they only sometimes require escaping. I have updated the PR to always escape them.

**Review by silentbicycle — COMMENTED** [review]
This looks good to me. I requested adding a test for the `?i-i` edge case, and there a few other potential test inputs on the PCRE pattern spec page (https://pcre.org/original/doc/html/pcrepattern.html#SEC13), but in the next couple weeks I also plan on doing comparative testing against pcregrep with generated regexes (using the subset of behavior we expect libfsm to handle), so that will shake out other unknown edge cases we need to test.

File: src/libre/parser.act [hunk 3]
```diff
@@ -705,35 +700,45 @@
...
+		@flags = *flags;
+	@};
+
+	<ast-set-re-flags>: (flags :re_flags) -> () = @{
+		*flags = @flags;
+	@};
+
+	<ast-mask-re-flags>: (pos :re_flags, neg :re_flags) -> () = @{
+		/*
+		 * Note: in cases like `(?i-i)`, the negative is
```
[the remark this answers is not in the document]
Comment (hvdijk) [reply src/libre/parser.act:719]: Sure. I added `tests/pcre-flags/in5.re` containing `(?i-i)a`.

**Comment by katef** [comment]
We'll need to handle the new flags field during AST rewriting, if we introduce any new nodes. Looking at ast_rewrite.c at the moment, I don't think we currently construct any new nodes, only removing existing ones.

**Comment by katef** [comment]
Thank you again!
